selinux: Ignore pods with Recursive policy

jsafrane opened the pull request:
The SELinuxWarning controller should not report SELinux label conflicts on Pods that explicitly opted into "seLinuxChangePolicy: Recursive" . It will only report a conflict with other Pods using the same volume with "seLinuxChangePolicy: Mount" (or nil).

This brings the controller behavior in sync with Kubelet.

User has opted out from mounting volume with `mount -o context=<selinux label>` and wants the previous kubelet / CRI behavior (recursive relabeling of the whole volume). In all previous Kubernetes releases, this situation never caused any events or errors, so the new controller should behave the same.

The controller now emits warnings only on pods with `seLinuxChangePolicy: MountOption` or `nil`.

KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling
Enhancement issue: https://github.com/kubernetes/enhancements/issues/1710

```relase-note
NONE
```

jsafrane committed:
selinux: Ignore pods with Recursive policy

Pod that explicitly opted into "seLinuxChangePolicy: Recursive" should not
report conflicts with another SELinux labels. They will only report a
conflict with other Pods using the same volume with "seLinuxChangePolicy:
Mount" (or nil).

jsafrane committed:
Update tests

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-selinux

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-selinux

jsafrane commented:
/retest
> failed to get metrics from KCM

I blame pod network. There is nothing interesting in KCM logs (it did not panic, it was not restarted)

gnufied commented:
/assign

gnufied commented:
But this still locks out other pod from reading/writing to the volume if selinux labels were different. We just don't generate the metrics now?

jsafrane commented:
Yes, you can see all the changes are in the controller. Kubelet will allow / reject the pods as usual.

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-selinux

jsafrane committed:
Improve error reporting of KCM metrics

Show what is a test waiting for and make sure the metric name + label
appears in the timeout error.

jsafrane commented:
I found that the test error was not a flake, just the error message was really bad.
Added a commit that turns 
` [FAILED] failed to get metrics from KCM: error waiting for kube-controller-manager to expose metrics: context deadline exceeded; client rate limiter Wait returned an error: context deadline exceede`
into
`[FAILED] failed to get metrics from KCM: timed out waiting for KCM metrics selinux_warning_controller_selinux_volume_conflict{[pod1_name="pvc-volume-tester-l8nbf" pod2_name="pvc-volume-tester-gdmbc" pod1_namespace="csi-mock-volumes-selinux-metrics-7538" pod2_namespace="csi-mock-volumes-selinux-metrics-7538" property="SELinuxLabel"]}`

Running the e2e tests **expecting** they fail to check the error is better

/test pull-kubernetes-e2e-gce-storage-selinux
/hold

jsafrane commented:
Ok, [this job run](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130472/pull-kubernetes-e2e-gce-storage-selinux/1895407851443064832) failed and logs report something useful:

> I0228 10:32:55.919440 76684 csi_selinux_mount.go:852] Waiting for KCM metric selinux_warning_controller_selinux_volume_conflict{[pod1_name="pvc-volume-tester-pbhg6" pod2_name="pvc-volume-tester-n5crm" pod1_namespace="csi-mock-volumes-selinux-metrics-8578" pod2_namespace="csi-mock-volumes-selinux-metrics-8578" property="SELinuxLabel"]}

...

> I0228 10:34:55.920188 76684 csi_selinux_mount.go:720] Unexpected error: failed to get metrics from KCM: 
    <*errors.errorString | 0xc00e085a80>: 
    error waiting for kube-controller-manager to expose metrics: context deadline exceeded; client rate limiter Wait returned an error: context deadline exceeded
    {
        s: "error waiting for kube-controller-manager to expose metrics: context deadline exceeded; client rate limiter Wait returned an error: context deadline exceeded",
    }

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-selinux

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-selinux

jsafrane commented:
Final message from [this run](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130472/pull-kubernetes-e2e-gce-storage-selinux/1895461829967089664): 

> { failed [FAILED] while waiting for metrics from KCM: error waiting for KCM metrics selinux_warning_controller_selinux_volume_conflict{[pod1_name="pvc-volume-tester-24zbd" pod2_name="pvc-volume-tester-vcllm" pod1_namespace="csi-mock-volumes-selinux-metrics-2164" pod2_namespace="csi-mock-volumes-selinux-metrics-2164" property="SELinuxLabel"]}: error waiting for kube-controller-manager to expose metrics: context deadline exceeded; client rate limiter Wait returned an error: context deadline exceeded
In [It] at: k8s.io/kubernetes/test/e2e/storage/csimock/csi_selinux_mount.go:720 @ 02/28/25 14:02:55.286

Looks good to me.

Pushed e2e test changes, they should pass now

/hold cancel.

gnufied commented:
/lgtm

jsafrane commented:
/hold cancel
/kind feature 
/release-note-none
